zcash_primitives/src/sapling/note.rs

```diff
@@ -0,0 +1,146 @@
+use group::{
+    ff::{Field, PrimeField},
+    GroupEncoding,
+};
+use rand_core::{CryptoRng, RngCore};
+
+use super::{value::NoteValue, Node, Nullifier, NullifierDerivingKey};
+use crate::keys::prf_expand;
+
+mod commitment;
+pub use self::commitment::{ExtractedNoteCommitment, NoteCommitment};
+
+pub(super) mod nullifier;
+
+/// Enum for note randomness before and after [ZIP 212](https://zips.z.cash/zip-0212).
+///
+/// Before ZIP 212, the note commitment trapdoor `rcm` must be a scalar value.
+/// After ZIP 212, the note randomness `rseed` is a 32-byte sequence, used to derive
+/// both the note commitment trapdoor `rcm` and the ephemeral private key `esk`.
+#[derive(Copy, Clone, Debug)]
+pub enum Rseed {
+    BeforeZip212(jubjub::Fr),
+    AfterZip212([u8; 32]),
+}
+
+impl Rseed {
+    /// Defined in [Zcash Protocol Spec § 4.7.2: Sending Notes (Sapling)][saplingsend].
+    ///
+    /// [saplingsend]: https://zips.z.cash/protocol/protocol.pdf#saplingsend
+    pub(crate) fn rcm(&self) -> commitment::NoteCommitTrapdoor {
+        commitment::NoteCommitTrapdoor(match self {
+            Rseed::BeforeZip212(rcm) => *rcm,
+            Rseed::AfterZip212(rseed) => {
+                jubjub::Fr::from_bytes_wide(prf_expand(rseed, &[0x04]).as_array())
+            }
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Note {
+    /// The value of the note
+    pub value: u64,
+    /// The diversified base of the address, GH(d)
+    pub g_d: jubjub::SubgroupPoint,
+    /// The public key of the address, g_d^ivk
+    pub pk_d: jubjub::SubgroupPoint,
+    /// rseed
+    pub rseed: Rseed,
+}
+
+impl PartialEq for Note {
+    fn eq(&self, other: &Self) -> bool {
+        self.value == other.value
+            && self.g_d == other.g_d
+            && self.pk_d == other.pk_d
+            && self.rcm() == other.rcm()
```

Comment:
Pre-existing issue: if, for the sake of argument, collisions could be found for $\mathsf{rseed} \mapsto \mathsf{PRF^{expand}_{rseed}}([4])$, then there could be two distinct `Note` instances that are equal. This is because `Note` actually represents a Sapling *note plaintext* minus the memo, not a Sapling note $(\mathsf{v}, \mathsf{d}, \mathsf{pk_d}, \mathsf{rcm})$ as defined in the spec.

(The protocol spec does not specify collision-resistance on the key as a security property of $\mathsf{PRF^{expand}_{rseed}}$. We can infer collision-resistance of its instantiation from that of BLAKE2b-512, so this is not a problem in practice.)